### PR TITLE
*: keep in-memory resident for model.TableInfo with special attributes

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: ref #50959 

Problem Summary:

`SchemaTables()` is a big problem for infoschema v2. It's slow, it cost a lot CPU resource, it make tidb memory thrash ... 

### What changed and how does it work?

I observe the pattern of call SchemaTables() API in our current codebase look-alike.
For ttl, tiflash and so on, they always iterate the SchemaTables and filter out some attributes.

```
	for _, dbName := range is.AllSchemaNames() {
		for _, tblInfo := range is.SchemaTableInfos(dbName) {
			if tblInfo.State != model.StatePublic || tblInfo.TTLInfo == nil {
				continue
			}
			...
		}
	}
```

So, **what if we keep all the model.TableInfo with special attributes in-memory?** 
This could be a temporary solution, dirty and quick but easy to implement and verify.

In this commit, I add a new `ListTablesWithSpecialAttribute()` api to replace `SchemaTableInfos`.
All the data are in memory, so it is as fast as or even faster than infoschema v1.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

The same data, memory usage after vs before:

image

We can see some jitters even with this change, it is caused by the statistics calling in this loop:

https://github.com/pingcap/tidb/blob/b9b330a9b6becafaa9be765e57726cb2661885a1/pkg/statistics/handle/bootstrap.go#L88-L109

And this commit only handle the memory in static state, when there are operations like `show xxx` or  `INFORMATION_SCHEMA.TABLES` visiting, there may still be `SchemaTables()` calling cause memory thrash. I'll handle them later.

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by D3Hunter** [comment]
> So, what if we keep all the model.TableInfo with special attributes in-memory?

can we only store those special attributes, not the whole table-info, as we have discussed in the coference?

**Comment by tiancaiamao** [comment]
> > So, what if we keep all the model.TableInfo with special attributes in-memory?
> 
> can we only store those special attributes, not the whole table-info, as we have discussed in the coference?

The callee still need the table info. For example, you can check here
https://github.com/pingcap/tidb/blob/3ad726d34cdf25e556c049b699f0bddb894a8e84/pkg/ddl/ddl_tiflash_api.go#L274-L294


Or do you mean just the table ID rather than *model.TableInfo? the pointer size itself is just a machine word.

@ywqzzy

**Comment by wjhuang2016** [comment]
> > > So, what if we keep all the model.TableInfo with special attributes in-memory?
> > 
> > 
> > can we only store those special attributes, not the whole table-info, as we have discussed in the coference?
> 
> The callee still need the table info. For example, you can check here
> 
> https://github.com/pingcap/tidb/blob/3ad726d34cdf25e556c049b699f0bddb894a8e84/pkg/ddl/ddl_tiflash_api.go#L274-L294
> 
> Or do you mean just the table ID rather than *model.TableInfo? the pointer size itself is just a machine word.
> 
> @ywqzzy

Store these infos may consume too much memory.

**Comment by tiancaiamao** [comment]
> > > > So, what if we keep all the model.TableInfo with special attributes in-memory?
> > > 
> > > 
> > > can we only store those special attributes, not the whole table-info, as we have discussed in the coference?
> > 
> > 
> > The callee still need the table info. For example, you can check here
> > https://github.com/pingcap/tidb/blob/3ad726d34cdf25e556c049b699f0bddb894a8e84/pkg/ddl/ddl_tiflash_api.go#L274-L294
> > 
> > Or do you mean just the table ID rather than *model.TableInfo? the pointer size itself is just a machine word.
> > @ywqzzy
> 
> Store these infos may consume too much memory.

If we don't store them, the we go for the Data Dictionary solution. @wjhuang2016

File: pkg/infoschema/infoschema.go [hunk 1]
```diff
@@ -320,6 +320,29 @@ func (is *infoSchema) SchemaTableInfos(schema model.CIStr) []*model.TableInfo {
 	return getTableInfoList(is.SchemaTables(schema))
 }
 
+type tableInfoResult struct {
+	DBName    string
+	TableInfo []*model.TableInfo
+}
+
+func (is *infoSchema) ListTablesWithSpecialAttribute(filter specialAttributeFilter) <-chan tableInfoResult {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply pkg/infoschema/infoschema.go:328]: Not now. If we need a ctx, then `is.SchemaTableInfos` etc all need it, and a lot other codes.

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/infoschema/infoschema.go [hunk 2]
```diff
@@ -320,6 +320,29 @@ func (is *infoSchema) SchemaTableInfos(schema model.CIStr) []*model.TableInfo {
 	return getTableInfoList(is.SchemaTables(schema))
 }
 
+type tableInfoResult struct {
+	DBName    string
+	TableInfo []*model.TableInfo
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema.go:325]:
```suggestion
	TableInfos []*model.TableInfo
```

File: pkg/infoschema/infoschema_v2.go [hunk 3]
```diff
@@ -171,11 +188,15 @@ func (isd *Data) add(item tableItem, tbl table.Table) {
...
 	isd.tableCache.Set(tableCacheKey{item.tableID, item.schemaVersion}, tbl)
-	if pi := tbl.Meta().GetPartitionInfo(); pi != nil {
+	ti := tbl.Meta()
+	if pi := ti.GetPartitionInfo(); pi != nil {
 		for _, def := range pi.Definitions {
 			isd.pid2tid.Set(partitionItem{def.ID, item.schemaVersion, tbl.Meta().ID, false})
 		}
 	}
+	if hasSpecialAttributes(ti) {
+		isd.tableInfoResident.Set(tableInfoItem{item.dbName, item.tableID, item.schemaVersion, ti, false})
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:198]: prefer we use named style like below, so many fields
```
tableID: xxx,
```

File: pkg/infoschema/infoschema_v2.go [hunk 4]
```diff
@@ -191,6 +212,7 @@ func (isd *Data) remove(item tableItem) {
 	item.tomb = true
 	isd.byID.Set(item)
 	isd.byName.Set(item)
+	isd.tableInfoResident.Set(tableInfoItem{item.dbName, item.tableID, item.schemaVersion, nil, true})
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:215]: doitto

File: pkg/infoschema/infoschema.go (hunk 1 above)
Comment (D3Hunter) [inline pkg/infoschema/infoschema.go:328]:
```suggestion
func (is *infoSchema) ListTablesWithSpecialAttribute(dbName string, filter specialAttributeFilter) <-chan tableInfoResult {
```
can we change it to get all tables with matched attributes for a db? so we don't to have a routine to do it. it's more complex now

Comment (tiancaiamao) [reply pkg/infoschema/infoschema.go:328]: This makes the streaming API possible. For example, if we don't store the special attribute and need to fetch data from tikv, then return all data []XXX may use much more memory then this API setting. @D3Hunter

**Review by D3Hunter — APPROVED** [review]

**Comment by D3Hunter** [comment]
/hold

please fix the callsite of the api, you can unhold after change

**Pull request merged**